### PR TITLE
Improvement/sql migrator

### PR DESCRIPTION
**Pull request by DylanTinianov**
Closes: #60 

## Description

Created a migrator to migrate models from Google datastore to the new Postgres SQL database.
______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/onflow/flow-playground/blob/master/CONTRIBUTING.md))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [standards mentioned here](https://github.com/onflow/flow-playground/blob/master/CONTRIBUTING.md).
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: migrate/apiVersion/migrate.go [hunk 1]
```diff
@@ -16,7 +16,7 @@
  * limitations under the License.
  */
 
-package migrate
+package apiVersion
```
Comment (sideninja) [inline migrate/apiVersion/migrate.go:19]: I wouldn't rename this package. I don't think "apiVersion" explains the functionality. I would rather keep it in the top level as it was before and just put the SQL migration to a separate package called "database" so it becomes "migrate/database"

File: migrate/sqlMigrator/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
+version: "3"
+
+services:
+  datastore_emulator:
```
Comment (sideninja) [inline migrate/sqlMigrator/docker-compose.yml:4]: why was this added?

File: migrate/sqlMigrator/sqlMigrator.go [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+	sqlModel "github.com/dapperlabs/flow-playground-api/model"
+	"github.com/dapperlabs/flow-playground-api/storage"
+	"github.com/google/uuid"
+	"time"
+)
+
+// numErrors counts the errors that occur during migration
+var numErrors = 0
+
+func main() {
```
Comment (sideninja) [inline migrate/sqlMigrator/sqlMigrator.go:18]: I would put main in "migrate/database/cmd/main.go" as per convention

File: migrate/sqlMigrator/sqlMigrator.go [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+		if !proj.Persist {
+			continue
+		}
+		migrateProject(dstore, sqlDB, proj) // Includes transaction & script execution templates
+		migrateAccounts(dstore, sqlDB, proj.ID)
+		migrateUser(dstore, sqlDB, proj)
+		migrateScriptExecutions(dstore, sqlDB, proj.ID)
+		migrateTransactionExecutions(dstore, sqlDB, proj.ID)
+	}
+	fmt.Println("Migration finished with", numErrors, "errors")
```
Comment (sideninja) [inline migrate/sqlMigrator/sqlMigrator.go:35]: I would create a logger instead of using fmt.

File: migrate/sqlMigrator/sqlMigrator.go [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+		Password: "password",
+		Name:     "postgres",
+		Port:     5432,
+	})
+	fmt.Println("Connected to SQL database")
+	return sqlDB
+}
+
+// getAllProjects returns a list of all projects in the datastore
+func getAllProjects(store *datastore.Datastore) *[]*model.InternalProject {
```
Comment (sideninja) [inline migrate/sqlMigrator/sqlMigrator.go:64]: You should paginate, this will become too much data that will need to be held in memory, possibly reaching the limit

Comment (sideninja) [reply migrate/sqlMigrator/sqlMigrator.go:64]: You can even create a wrapper function that knows how to handle pagination and then no matter which resource you are getting just make sure you are getting it in batch until the result is null which should indicate no more data.

File: migrate/sqlMigrator/sqlMigrator.go [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+		fmt.Println(err)
+		panic(err)
+	}
+	return &projects
+}
+
+// migrateAccounts migrates models of datastore accounts to sql accounts
+func migrateAccounts(dstore *datastore.Datastore, sqlDB *storage.SQL, projID uuid.UUID) {
+	var accounts []*model.InternalAccount
+	err := dstore.GetAccountsForProject(projID, &accounts)
```
Comment (sideninja) [inline migrate/sqlMigrator/sqlMigrator.go:78]: this is ok to be fetched in a single query, not a lot of accounts

File: migrate/sqlMigrator/sqlMigrator.go [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+	if err != nil {
+		numErrors++
+		return
+	}
+
+	var sqlAccounts []*sqlModel.Account
```
Comment (sideninja) [inline migrate/sqlMigrator/sqlMigrator.go:84]: better for performance
```suggestion
	sqlAccounts := make([]*sqlModel.Account, 0, len(accounts))
```

File: migrate/sqlMigrator/sqlMigrator.go [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+	for _, acc := range accounts {
+		tmp := sqlModel.Account{}
+		tmp.ID = acc.ProjectChildID.ID
+		tmp.ProjectID = acc.ProjectChildID.ProjectID
+		tmp.Address = sqlModel.Address(acc.Address)
+		tmp.Index = acc.Index
+		tmp.DraftCode = acc.DraftCode
+		sqlAccounts = append(sqlAccounts, &tmp)
```
Comment (sideninja) [inline migrate/sqlMigrator/sqlMigrator.go:92]:
```suggestion
		sqlAccounts[i] = &tmp
```

File: migrate/sqlMigrator/sqlMigrator.go [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+	}
+
+	var sqlAccounts []*sqlModel.Account
+	for _, acc := range accounts {
+		tmp := sqlModel.Account{}
+		tmp.ID = acc.ProjectChildID.ID
+		tmp.ProjectID = acc.ProjectChildID.ProjectID
+		tmp.Address = sqlModel.Address(acc.Address)
+		tmp.Index = acc.Index
+		tmp.DraftCode = acc.DraftCode
```
Comment (sideninja) [inline migrate/sqlMigrator/sqlMigrator.go:91]: Why not just
```suggestion
		sqlAccounts[i] = sqlModel.Account{
		     ID: acc.ProjectChildID.ID,
		     // ...
		}
```

File: migrate/sqlMigrator/sqlMigrator.go [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+	err := dstore.GetTransactionTemplatesForProject(projID, &ttpl)
+	if err != nil {
+		fmt.Println(err)
+		fmt.Println("Error: could retrieve transaction templates for project ID", projID,
+			"skipping project")
+		numErrors++
+		return nil
+	}
+
+	var sqlTtpl []*sqlModel.TransactionTemplate
```
Comment (sideninja) [inline migrate/sqlMigrator/sqlMigrator.go:115]: same as above comment

File: migrate/sqlMigrator/sqlMigrator.go [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+// migrateScriptTemplates converts models of datastore script templates to sql script templates
+func convertScriptTemplates(dstore *datastore.Datastore, projID uuid.UUID) *[]*sqlModel.ScriptTemplate {
+	var stpl []*model.ScriptTemplate
+	err := dstore.GetScriptTemplatesForProject(projID, &stpl)
+	if err != nil {
+		fmt.Println(err)
+		return nil
+	}
+
+	var sqlStpl []*sqlModel.ScriptTemplate
```
Comment (sideninja) [inline migrate/sqlMigrator/sqlMigrator.go:137]: same as above

File: migrate/sqlMigrator/sqlMigrator.go [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+		tmp.ProjectID = acc.ProjectChildID.ProjectID
+		tmp.Address = sqlModel.Address(acc.Address)
+		tmp.Index = acc.Index
+		tmp.DraftCode = acc.DraftCode
+		sqlAccounts = append(sqlAccounts, &tmp)
+	}
+
+	err = sqlDB.InsertAccounts(sqlAccounts)
+	if err != nil {
+		fmt.Println("Error on migrate accounts for project ID", projID.String(), err)
```
Comment (sideninja) [inline migrate/sqlMigrator/sqlMigrator.go:97]: It would be good to output more account information, best would be the whole account so we can see whether there's something weird going on

File: migrate/sqlMigrator/sqlMigrator.go [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+		tmp.Title = ttp.Title
+		tmp.Index = ttp.Index
+		tmp.Script = ttp.Script
+		sqlStpl = append(sqlStpl, &tmp)
+	}
+	return &sqlStpl
+}
+
+func convertProject(proj *model.InternalProject) *sqlModel.Project {
+	sqlProj := &sqlModel.Project{}
```
Comment (sideninja) [inline migrate/sqlMigrator/sqlMigrator.go:151]: inline create as above

File: migrate/sqlMigrator/sqlMigrator_test.go [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+
+func generateProject(projectGenCount int) (*model.InternalProject, *[]*model.TransactionTemplate, *[]*model.ScriptTemplate) {
+	proj := &model.InternalProject{
+		ID:                        uuid.New(),
+		UserID:                    uuid.New(),
+		Secret:                    uuid.New(),
+		PublicID:                  uuid.New(),
+		ParentID:                  nil,
+		Title:                     "Project number " + strconv.Itoa(projectGenCount),
+		Description:               "",
```
Comment (sideninja) [inline migrate/sqlMigrator/sqlMigrator_test.go:74]: add values here

File: migrate/sqlMigrator/sqlMigrator_test.go [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+
+// generateAccounts generates a random number of accounts up to 10
+func generateAccounts(projID uuid.UUID) *[]*model.InternalAccount {
+	var accounts []*model.InternalAccount
+
+	for i := 0; i < rand.Intn(10); i++ {
+		// Generate address
+		bn := make([]byte, 8)
+		binary.BigEndian.PutUint32(bn, uint32(i))
+		addr := model.NewAddressFromBytes(bn)
```
Comment (sideninja) [inline migrate/sqlMigrator/sqlMigrator_test.go:130]: All accounts in the projects are just "0x01", "0x02", "0x03", "0x04", so just replicate that in tests

File: migrate/sqlMigrator/sqlMigrator_test.go [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+	var scriptExecs []model.ScriptExecution
+	for i := 0; i < proj.TransactionExecutionCount; i++ {
+		scriptExecs = append(scriptExecs, model.ScriptExecution{
+			ProjectChildID: model.ProjectChildID{
+				ID:        uuid.New(),
+				ProjectID: proj.ID,
+			},
+			Index:     i,
+			Script:    "Test script execution",
+			Arguments: nil,
```
Comment (sideninja) [inline migrate/sqlMigrator/sqlMigrator_test.go:156]: lack of arguments, errors, and logs is problematic if you want just to use tests to prove functioning of this migration. Those values would in turn be migrated.

File: migrate/sqlMigrator/sqlMigrator_test.go [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+	var txExecs []model.TransactionExecution
+	for i := 0; i < proj.TransactionExecutionCount; i++ {
+		txExecs = append(txExecs, model.TransactionExecution{
+			ProjectChildID: model.ProjectChildID{
+				ID:        uuid.New(),
+				ProjectID: proj.ID,
+			},
+			Index:     i,
+			Script:    "Test transaction execution",
+			Arguments: nil,
```
Comment (sideninja) [inline migrate/sqlMigrator/sqlMigrator_test.go:175]: same as above

File: migrate/sqlMigrator/docker-compose.yml (hunk 2 above)
Comment (DylanTinianov) [reply migrate/sqlMigrator/docker-compose.yml:4]: Added for the datastore viewer as it's sort of required for testing. I can remove it from the project if you think it's not necessary, and maybe leave a comment about the datastore viewer somewhere.
